### PR TITLE
feat(api): XIP-83 d14n binding — bidirectional Subscribe on QueryApi

### PR DESCRIPTION
**Pull request by tylerhawkes**
## XIP-83 d14n binding — bidirectional `QueryApi.Subscribe`

Implements the decentralized (d14n) binding of [XIP-83 mutable subscription streams](https://github.com/xmtp/XIPs/pull/139) on xmtpd. A single long-lived bidirectional stream lets a client **mutate its topic set in place** (add/remove subscriptions) and keep the connection alive with ping/pong — instead of tearing down and reopening a stream every time group membership changes.

This is the xmtpd counterpart to the node-go v3 `MlsApi.Subscribe` binding. Same control protocol (`Mutate` / `Started` / `CatchupComplete` / `TopicsLive` / `Ping` / `Pong`); the d14n binding delivers `OriginatorEnvelope`s and uses per-originator vector cursors.

### Design: gated async catch-up
- **Single-writer + single-sender actor** per subscription. All topic-set mutation happens on the writer goroutine; all stream sends happen on the sender goroutine (drained via a bounded queue).
- **Per-topic gate + pending buffer.** A newly-added topic is caught up by an async fetcher goroutine while already-live topics keep flowing — a slow catch-up never head-of-line-blocks live delivery.
- **No missing messages across the switch.** A topic is registered *live* **before** its catch-up snapshot is taken, and delivery is deduped by a monotonic per-originator cursor (`advanceTopicCursors`). Anything that lands during catch-up is either in the snapshot or arrives live; the dedup drops the overlap.
- **Sparse vs. filled cursors.** The writer keeps only the originator cursors the client provided (growing them as new originators are seen); the fetcher holds a filled copy purely for the query. Keeps memory bounded at the active-topic ceiling.
- **1M active-topic ceiling** (`maxActiveSubscribeTopics`) — ~16–32MB for a consumer that large, deliberately favored over forcing clients into multiple connections (which would invite rate-limiting and wreck some topologies).
- **Liveness.** Either peer may `Ping`; the receiver must `Pong` the nonce. No pong within the deadline → the node reaps the vanished peer (e.g. a mobile client the OS suspended behind a proxy that still ACKs the transport).

### Commits
1. `feat(api): add mutableSubscription handle to subscribeWorker` — in-place topic add/remove on the existing subscribe worker, guarded by `topicsMu`.
2. `feat(api): XIP-83 bidirectional Subscribe handler on QueryApi + xmtpv4 protos` — the handler (`subscribe.go`) + regenerated xmtpv4 protos.
3. `test(api): Subscribe handler tests + configurable keepalive interval for tests` — 5 race-tested scenarios + a test-only knob to shorten the keepalive cadence.

### Tests
`go test -race ./pkg/api/message` — all green:
- `TestSubscribe_CatchUpThenLive`
- `TestSubscribe_MutateRemoveStopsDelivery`
- `TestSubscribe_HalfCloseHistoryOnlyDrains`
- `TestSubscribe_HistoryOnlyOnLiveRejected`
- `TestSubscribe_NoPongIsReaped`

### Dependency & merge ordering
Built on [xmtp/proto#338](https://github.com/xmtp/proto/pull/338) (`QueryApi.Subscribe`). The committed `.pb.go` here is generated from that branch via `dev/gen/protos`; once #338 merges, regenerating from proto `main` produces byte-identical output.

> Note: proto#338's `Test (Golang)` check cross-compiles xmtpd `main` against the new proto and will stay red until this handler lands on `main` (xmtpd's `*Service` must implement `Subscribe`). Recommended order: merge proto#338 (its red Golang check is expected), then this.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

> [!NOTE]
> ### Add bidirectional Subscribe RPC to QueryApi with mutable topic membership and catch-up history
> - Implements the XIP-83 `Subscribe` bidi-streaming RPC in [`pkg/api/message/subscribe.go`](https://github.com/xmtp/xmtpd/pull/2020/files#diff-b73bd7f55e5ea34b032c9b6416652d6d4fec75355fc8dd0ec53c91d414bd5a0e): clients send `Mutate` frames to add/remove topics at runtime; the server delivers history catch-up, `TopicsLive`, and `CatchupComplete` markers per mutate, then routes live envelopes.
> - Introduces `mutableSubscription` in [`pkg/api/message/mutable_subscription.go`](https://github.com/xmtp/xmtpd/pull/2020/files#diff-f0062837aafd15807b4f6c75075ab2bea151bdf495808d3c3ecedaafe400c851) so topic registration is updated in-place without reconnecting; add/remove are idempotent and safe under concurrent worker reaping.
> - Adds `subscribeSession` state machine managing ping/pong keepalive, backpressure-bounded sending, pending-byte caps, in-flight wave limits, and graceful half-close drain.
> - Fixes `closeListener` in [`pkg/api/message/subscribe_worker.go`](https://github.com/xmtp/xmtpd/pull/2020/files#diff-6ead5de931f1e9d71c83166064bf5ca1a0d1d7bbeab213d8d26cb0781237a5ab) to set the closed flag under `topicsMu` before channel close, preventing re-registration races and send-on-closed-channel bugs.
> - Adds generated protobuf, gRPC, and Connect-Go stubs for `SubscribeRequest`/`SubscribeResponse` and related message types across `pkg/proto/`.
>
> 
>
> <sup><a href="https://app.macroscope.com">Macroscope</a> summarized c44972f.</sup>
> 
>

### CODE REVIEW

File: pkg/api/message/mutable_subscription.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	m.l.topicsMu.Lock()
+	defer m.l.topicsMu.Unlock()
+	if m.l.closed {
+		return
+	}
+	m.worker.topicListeners.addListener(keys, m.l)
+	for k := range keys {
+		m.l.topics[k] = struct{}{}
+	}
+}
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/mutable_subscription.go:56]: Fixed. `closeListener` now sets `l.closed` **under `l.topicsMu`**, so a concurrent `mutableSubscription.addTopics` observes it atomically with its `addListener` call and can no longer re-register a listener after its channel is closed. Added a `-race` regression test (`TestMutableSubscriptionAddRaceWithClose`) that hammers addTopics against the reap.

File: pkg/api/message/subscribe.go [hunk 2]
```diff
@@ -0,0 +1,685 @@
...
+		queryCursors,
+		cursorKeys,
+		sess.catchUpCh,
+		sess.logger,
+	)
+	sess.nextWave++
+	return nil
+}
+
+func (sess *subscribeSession) removeTopic(parsed *topic.Topic) {
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go:871]: Fixed. Each gated topic now records the wave currently catching it up — the gate map became `catchingUp map[cursorKey]int` (presence = gated, value = owning wave; it replaced the old bool gate + a parallel map). A wave only flushes/opens/announces topics it still owns (`catchingUp[t] == wave`) at completion, and a stale wave's in-flight pages for a removed/re-added topic are dropped (`envsOwnedByWave`). So a remove+re-add (reset) while a prior wave is still fetching no longer prematurely opens the new wave's gate. Regression test `TestSubscribeSessionStaleWaveSkipsResetTopic`.

File: pkg/api/message/subscribe.go [hunk 3]
```diff
@@ -0,0 +1,685 @@
...
+	knownOriginators, err := sess.svc.originatorList.GetOriginatorNodeIDs(sess.ctx)
+	if err != nil {
+		return connect.NewError(
+			connect.CodeInternal,
+			fmt.Errorf("could not get originator list: %w", err),
+		)
+	}
+	historyOnly := m.GetHistoryOnly()
+
+	for _, w := range m.GetRemoves() {
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go:671]: Fixed. `handleMutate` now **fully validates before mutating any state**: it parses all removes and adds, dedups, enforces the history_only / overlapping-catch-up rules, and checks the active-topic cap (via a projected-size delta) up front, then applies removes-then-adds only once validation passes. (A malformed Mutate already tore the whole stream down, so there was no observable partial state in practice, but the validation is now atomic regardless.)

File: pkg/api/message/subscribe.go [hunk 4]
```diff
@@ -0,0 +1,685 @@
...
+	}
+
+	// Wave complete: for each live topic still subscribed, flush its gated live (deduped against
+	// the now-advanced cursor) and open its gate; announce the live topics; then CatchupComplete.
+	wire := make([][]byte, 0, len(w.topics))
+	for _, t := range w.topics {
+		if w.historyOnly {
+			wire = append(wire, t.wire)
+			continue
+		}
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go:586]: Working as specified, leaving as-is. The proto contract for the `history_only` field explicitly mandates these markers — it says the adds are caught up via *"history, TopicsLive markers, and the wave's CatchupComplete"* — and `TopicsLive` is documented informational-only (*"delivery correctness (no duplicates, no gaps) never depends on it"*). A client that opted into `history_only` interprets `TopicsLive` as the "everything as of now" boundary, not "a live tail follows"; it learns there is no live tail because it sent `history_only` (and typically half-closes). Suppressing the marker here would make this binding diverge from the field contract and from the v3 MlsApi binding.

**Comment by tylerhawkes** [comment]
Pushed an update addressing the Macroscope findings (3 fixed, 1 working-as-specified per the `history_only` proto contract — see inline replies).

Beyond those, I did a deeper self-review of the concurrent paths and hardened several edge cases, each with a regression test:

- **Liveness**: decoupled the pong-reap deadline from the delivery/ping cadence (a single shared ticker meant steady outbound traffic could postpone the reap indefinitely; long half-close drains could be falsely reaped). Ping cadence now tracks real sends; the reap deadline is independent and suppressed during half-close.
- **Re-add**: a plain re-add of an already-live topic (no remove) is now a no-op instead of resetting the cursor and replaying — replay still requires remove+re-add.
- **Teardown**: `flush()` now propagates a sender send-error instead of returning a false OK with the wave's terminal frames undelivered; the stream context is cancelable so catch-up fetchers tear down promptly.
- **Writer latency**: the per-Mutate originator-list lookup (a DB round-trip on a cache miss) moved off the single writer goroutine into the catch-up fetcher, so a slow DB can no longer stall liveness or false-reap a healthy stream.
- **Overlap**: rejecting a second overlapping catch-up for the same topic (including via remove+re-add over an in-flight `history_only` wave).

`go test -race ./pkg/api/message` is green; golangci-lint clean.

File: pkg/api/message/subscribe.go [hunk 5]
```diff
@@ -0,0 +1,869 @@
...
+				rowsPerEntry,
+			)
+			if err != nil {
+				emit(catchUpBatch{wave: wave, err: err})
+				return
+			}
+			envs := unmarshalEnvelopes(rows, logger)
+			// Advance the fetcher's own (filled) cursors so the next page continues; the proto
+			// result is discarded here — the writer re-dedups the raw envs against the live cursor.
+			_ = advanceTopicCursors(queryCursors, envs, logger)
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go]: Fixed. The fetcher now advances its pagination cursors from the **raw rows** (`advanceCursorsFromRows`), not just the successfully-unmarshaled envelopes — so a row whose envelope bytes fail to parse still moves the cursor past it and pagination always makes forward progress. The emitted envelopes are still re-deduped against the live cursor by the writer. Regression test `TestAdvanceCursorsFromRowsAdvancesPastUnmarshalFailures`.

File: pkg/api/message/subscribe.go [hunk 6]
```diff
@@ -0,0 +1,869 @@
...
+		// A topic with an in-flight history_only catch-up can't take a second overlapping catch-up
+		// (live or history_only) — even via remove+re-add: removeTopic does NOT cancel a history_only
+		// wave (it only clears live state), so both waves would paginate and deliver T's history.
+		if sess.hasInflightHistoryOnly(cursorKey) {
+			return connect.NewError(
+				connect.CodeInvalidArgument,
+				errors.New("add targets a topic with an in-flight history_only catch-up"),
+			)
+		}
+		provided := make(db.VectorClock)
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go:729]: Fixed. `handleMutate` now validates every `LastSeen` entry and rejects the Mutate with `InvalidArgument` if a node id exceeds `MaxInt32` or a sequence id exceeds `MaxInt64`, so an out-of-range cursor can no longer be silently dropped by `SetPerTopicCursors` or stored verbatim into the live cursor (which would have marked every real envelope from that originator as already-seen and killed the topic). Regression test `TestSubscribeSessionRejectsOutOfRangeCursor`.

File: pkg/api/message/subscribe.go [hunk 7]
```diff
@@ -0,0 +1,869 @@
...
+	emit := func(b catchUpBatch) bool {
+		select {
+		case catchUpCh <- b:
+			return true
+		case <-ctx.Done():
+			return false
+		}
+	}
+
+	knownOriginators, err := s.originatorList.GetOriginatorNodeIDs(ctx)
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go:952]: Documented as an accepted eventual-consistency property of `history_only` (comment added at the snapshot). For a **live** wave this is already complete: the listener is registered *before* the snapshot, and the worker dispatches by topic independently of this originator list, so any originator that first publishes after the snapshot is caught by the live gate and flushed at completion. A `history_only` wave has no live gate, so a brand-new originator that first publishes *during* the bounded catch-up is delivered on the client's next subscribe rather than this sync. Since `history_only` is the periodic re-sync flow, that residual window is acceptable. A stronger guarantee would mean registering a transient live gate for the history_only window (then discarding it at completion) — happy to add that if a use case needs it, but it adds dispatch load for a case the next sync already covers.

**Comment by tylerhawkes** [comment]
## Response to the Octane security review

Thanks — these were valuable. Note Octane scanned commit `c44b65c` (the first push); several findings were already fixed by a deep-review pass in the current commit, and I've now fixed the rest that are real. Triage of all 18:

### ✅ Fixed in this push (in direct response to Octane + Macroscope)
- **Per-send `time.After` allocation** → `send()` now reuses one per-session `*time.Timer` instead of allocating a timer (live for the whole keepalive) on every send.
- **Unbatched / size-unchecked flush** → `sendEnvelopes` now splits into ≤2 MiB frames, so a large catch-up page or a flushed pending buffer can never go out as one oversized, stream-aborting frame.
- **Cursor `int32`/`int64` overflow** (Macroscope too) → out-of-range `LastSeen` entries are now rejected with `InvalidArgument` instead of being silently dropped / poisoning the live cursor.
- **Catch-up spin on an unparseable row** (Macroscope too) → pagination cursors now advance from raw rows, guaranteeing forward progress.

### ✅ Already fixed in the current commit (Octane scanned the pre-fix `c44b65c`)
- **Data race on `listener.closed` + non-idempotent removal** → `closeListener` now sets `closed` under `topicsMu`; race test added.
- **Keepalive ping/pong not suspended after half-close** → ping send + reap are gated on `!halfClosed`.
- **Non-atomic mutate handling** → `handleMutate` fully validates before applying.
- **Lack of per-topic wave serialization / mixed dedup (dupes + out-of-order)** → each gated topic records its owning wave (`catchingUp map[key]int`); a stale wave can't open a reset topic's gate, and its in-flight pages are dropped.
- **Missing sender error check in `flush()`** → `flush()` now drains `sendErrCh` and propagates instead of a false OK.

### ⚖️ By design (with rationale)
- **1M live-topic budget + "missing rate limits"** → the 1M cap is intentional (~16–32 MB; a herald wants one fat connection, not fan-out that just hits per-connection limits). Request-rate is bounded by the existing API rate-limit interceptor; the per-topic memory ceiling and the 64 MiB pending cap are the resource bounds.
- **In-place "history replays" / cursor monotonicity** → a plain re-add of an already-live topic is now a no-op (no replay); replay requires explicit remove+re-add. So unmetered replay-by-re-add is gone.
- **Silent ignore of unknown request types** → forward-compat by design: unknown response/request features are gated behind the `Started.capabilities` advertisement (a client must not send an unadvertised type); an unknown *version* arm IS rejected with `InvalidArgument`.
- **Immediate per-topic removal drops in-flight envelopes** → intended: removing a topic means stop delivering it, including buffered/in-flight envelopes.
- **Half-close doesn't drain the live producer tail** → the half-close drain finishes in-flight catch-up waves then closes; a client that half-closed for a bounded sync isn't expecting a live tail.
- **removes-first ordering allows same-frame `history_only` reset** → the client explicitly asked for `history_only` on that topic; and remove+re-add over an *in-flight* history_only wave is now rejected (added this pass + test).

### 📋 Acknowledged — shared infra / deployment (out of scope for this PR, flagged for follow-up)
- **Write-locked O(N) listener removal** and **count-based `l.ch` buffering** live in the shared `subscribeWorker`/`listenersMap` dispatch layer (predates this PR); the mutable subscription exercises removal more, so a finer-grained or sharded listener map is a worthwhile follow-up.
- **`http.Server.ReadTimeout` force-closing long-lived streams** is a server-config concern that affects every streaming RPC (including the existing `SubscribeTopics`), not this handler — worth confirming the deploy config exempts bidi/streaming.
- **TTL-cached originator enumeration** — same eventual-consistency note as the Macroscope `history_only` thread; live waves are unaffected (topic-based gate).
- **Unbounded concurrent catch-up waves** — bounded today by the API request-rate limiter (each wave = one Mutate); a per-stream in-flight-wave cap would be reasonable defense-in-depth if we want it.

`go test -race ./pkg/api/message` is green; golangci-lint clean; new regression tests cover the cursor-overflow, catch-up-spin, frame-splitting, and the already-fixed concurrency/teardown items.

**Comment by tylerhawkes** [comment]
## Handler state-management redesigned to make the recurring bug-classes structurally impossible

Across the prior review rounds (Macroscope ×3 + a deep multi-agent pass), the findings kept clustering into the **same two structural causes** rather than independent bugs. Instead of patching symptoms again, I restructured the handler so those classes can't be expressed:

**1. Per-topic state was four parallel maps** (`catchingUp`, `cursors`, `liveTopics`, `pending`), kept consistent by hand at every call site — so each new code path was a fresh chance to update one and forget another (most of the recurring findings). Collapsed into **one `map[string]*topicState`**, a small `gated → live` state machine; every transition goes through a method (`gateTopic` / `bufferLive` / `flushAndGoLive` / `removeTopicState`) that moves the whole struct together. The parallel `waveCursors` map folded onto `subscribeWave.cursors` (history_only waves only).

**2. The sender result** was a `(sendErrCh, cap 1)` + `senderDone` pair with a non-blocking re-read in `flush` — the source of the false-OK class, plus a latent single-consumer race on the buffered channel. Collapsed into a **single broadcast-on-close contract** (`senderDone` + a `sendErr` field): the sender records its terminal status, then closes; every reader learns the outcome by observing the close and reading the field.

All previously-flagged fixes are folded in: `removeListener` idempotency, `flush` returning `CodeCanceled` on ctx-cancel, cursor-overflow rejection, frame-splitting, and the in-flight history_only overlap guard.

### Review

Two independent multi-agent review rounds (5 dimensions each; every finding adversarially verified by 3 skeptics). The `topicState` invariants, the new sender contract, the cap/`pendingBytes` arithmetic, and wave-ownership all got clean sign-offs. Confirmed issues across both rounds, all addressed:

- A **pre-existing teardown `Send`-race** — the bounded wait on `senderDone` abandons a sender wedged inside connect's non-cancelable `stream.Send` on a non-reading client, so `Close` can race the response writer. It's a connect-go limitation (an unbounded wait would hang the handler instead); documented honestly in the teardown comment and tracked separately.
- An **unbounded in-flight-waves DoS gap** (new, found by the structural review): the active-topic cap didn't bound concurrent catch-up fetcher goroutines, and `remove + re-add` churn nets zero. Fixed with a `maxInflightSubscribeWaves` cap + test.

Build clean, tests race-clean, golangci-lint v2.10.1: 0 issues. Note: the line anchors on the earlier inline comments have moved with the rewrite — replies inline.

File: pkg/api/message/subscribe.go [hunk 8]
```diff
@@ -0,0 +1,941 @@
...
+		frame = nil // do NOT reuse the backing array; the sent frame still references it
+		frameBytes = 0
+		return nil
+	}
+	for _, env := range envs {
+		size := proto.Size(env)
+		if len(frame) > 0 && frameBytes+size > limit {
+			if err := flush(); err != nil {
+				return err
+			}
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go:474]: Working as intended — not dropping. `maxSubscribeFrameBytes` (2 MiB) is a soft batching target, ~12× below the transport's hard cap (`GRPCPayloadLimit` = 25 MiB). An envelope larger than 2 MiB isn't skipped: it flushes the current frame and goes out alone, and it always fits, because it was publishable under that same 25 MiB cap. Skipping it (as the old `batchAndSendEnvelopes` did) would silently lose a valid, deliverable message — strictly worse. Documented inline at the `sendEnvelopes` loop. (The handler was since redesigned, so this line anchor has moved.)

File: pkg/api/message/mutable_subscription.go [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+	defer m.l.topicsMu.Unlock()
+	m.worker.topicListeners.removeListener(keys, m.l)
+	for k := range keys {
+		delete(m.l.topics, k)
+	}
+}
+
+// close unregisters the subscription from every topic it still holds. Safe to call even after
+// the worker has already reaped the listener (removeListener is idempotent).
+func (m *mutableSubscription) close() {
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/mutable_subscription.go:73]: Fixed. `listenersMap.removeListener` now `continue`s instead of `return`ing when a key is already gone, so a batched `close()` (or the worker's reap path, which also passes a multi-key map) that hits an already-reaped topic keeps unregistering the rest instead of leaking the remainder.

File: pkg/api/message/subscribe.go [hunk 10]
```diff
@@ -0,0 +1,941 @@
...
+		// The sender finished — but it may have stopped early on a Send error, leaving the wave's
+		// terminal TopicsLive/CatchupComplete unsent. Surface that rather than a false OK.
+		select {
+		case err := <-sess.sendErrCh:
+			return err
+		default:
+			return nil
+		}
+	case <-sess.ctx.Done():
+		return nil
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go]: Fixed. `flush()` returns `CodeCanceled` (wrapping the ctx error) on `ctx.Done()` instead of `nil`, matching the keepalive-timeout arm — so a drain interrupted before the wave's terminal `TopicsLive` / `CatchupComplete` reach the wire surfaces as an error, not a false OK. The whole sender-result path was also redesigned into a single broadcast-on-close contract (`senderDone` + a `sendErr` field), which removed the non-blocking `sendErrCh` re-read this finding was about.

File: pkg/api/message/subscribe.go [hunk 11]
```diff
@@ -0,0 +1,1069 @@
...
+	if !sess.outClosed {
+		sess.outClosed = true
+		close(sess.outbound)
+	}
+}
+
+// flush drains queued frames before a GRACEFUL completion so no tail is lost. Bounded by the
+// keepalive deadline / ctx; if that bound trips the drain did NOT finish, so it returns
+// DeadlineExceeded rather than a false OK (the bug the v3 binding's flush originally had).
+func (sess *subscribeSession) flush() error {
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go:361]: Fixed. `flush()` now gives `senderDone` deterministic priority: when the `ctx.Done()` or keepalive-timeout arms fire, it re-checks (non-blocking) whether the sender has actually drained via a new `senderDrained()` helper, and if so returns `sendErr` (nil on a clean drain) instead of `CodeCanceled`/`DeadlineExceeded`. So a graceful close where the sender finished in the same instant as cancellation/timeout is never reported as a spurious error. Covered by `TestSubscribeSessionFlushPrefersDrainedSenderOverCancel` (64 iterations against the pseudorandom select — fails without the fix).

File: pkg/api/message/subscribe.go [hunk 12]
```diff
@@ -0,0 +1,1069 @@
...
+// right before the pong-deadline reap so a Pong that landed in the buffer just as the deadline fired
+// (Go select picks randomly among ready cases) is still counted, instead of false-reaping a stream
+// that did answer. ch may be nil (half-closed), in which case this is a no-op.
+func (sess *subscribeSession) drainPendingRequests(ch <-chan *message_api.SubscribeRequest) error {
+	for {
+		select {
+		case req, open := <-ch:
+			if !open {
+				return nil
+			}
```
[the remark this answers is not in the document]
Comment (tylerhawkes) [reply pkg/api/message/subscribe.go:1040]: Fixed. `drainPendingRequests` now mirrors the main loop's EOF handling on a closed channel: it returns `CodeUnavailable` (wrapping `recvErr`) on a transport failure, and sets `sess.halfClosed = true` on a clean EOF. So when the client half-closes (or `Receive` fails) while a ping is outstanding and the pong-deadline fires first, the stream surfaces the real recv error or finishes the clean half-close instead of being reaped with a spurious `DeadlineExceeded`. Covered by `TestSubscribeSessionDrainPendingRequestsHandlesClosedChannel` (mutation-verified: reverting to `return nil` fails it).

**Review by mkysel — APPROVED** [review]
